crypto/x509/x509_vfy.c

```diff
@@ -824,7 +820,8 @@ static int check_trust(X509_STORE_CTX *ctx, int num_untrusted)
      */
     for (i = num_untrusted; i < num; i++) {
         x = sk_X509_value(ctx->chain, i);
-        trust = X509_check_trust(x, ctx->param->trust, 0);
+        trust = X509_check_trust(x, ctx->param->trust,
+                                 ctx->param->flags & X509_V_FLAG_PARTIAL_CHAIN);
```

Comment:
I see. This won't be hard to correct.

Reply:
Done.
BTW, it looks like one could simply reuse `X509_TRUST_OK_ANY_EKU` here, right?
Yet for now for clarity I've introduced `X509_TRUST_DO_PC_COMPAT`,

Reply:
My undestanding was that `X509_TRUST_OK_ANY_EKU` is to indicate how to handle the special "anyExtendedKeyUsage" KeyPurposeId, which has the very curious behavior of:
```
   Applications that require the presence of a
   particular purpose MAY reject certificates that include the
   anyExtendedKeyUsage OID but not the particular OID expected for the
   application.
```

Reply:
Interesting explanation, but I just had a closer look at the implementation and it turns out that the semantics of `X509_TRUST_OK_ANY_EKU` is very different:

First of all, the `anyExtendedKeyUsage` extended key usage OID is recognized by `x509v3_cache_extensions()`, setting `XKU_ANYEKU` in the cert structure, but apart from that `anyExtendedKeyUsage` entries are **ignored**!

Independently of that, if `X509_TRUST_OK_ANY_EKU` is set in the verification flags, it leads to `X509_TRUST_REJECTED` if the explicit rejection list in the auxiliary info f a (trusted) certificate contains `anyExtendedKeyUsage` (which makes the name `X509_TRUST_OK_ANY_EKU` rather misleading),
and it leads to `X509_TRUST_TRUSTED` if the explicit trust list  in the auxiliary info of a (trusted) certificate contains `anyExtendedKeyUsage`.

So in a nutshell, `X509_TRUST_OK_ANY_EKU` governs whether `anyExtendedKeyUsage` in any auxiliary trust info of a trust store element is recognized as trust wildcard trust or distrust wildcard, respectively.

Reply:
Isn't it a shame that there is so much unclarity even among experts what various verification options do?
The `XKU_ANYEKU` is mentioned in the documentation but the only thing stated is only said that it corresponds to `anyExtendedKeyUsage` - and even this is misleading because this (pseudo-)EKU entry gets ignored.
And `X509_TRUST_OK_ANY_EKU` is not even mentioned in the documentation.

Reply:
Thank you for digging into this more.

> So in a nutshell, X509_TRUST_OK_ANY_EKU governs whether anyExtendedKeyUsage in any auxiliary trust info of a trust store element is recognized as trust wildcard trust or distrust wildcard, respectively.

I am not sure whether I am matching things up as you intended with the "respectively".  In my reading, `X509_TRUST_OK_ANY_EKU` must be set in the flags in order for anyExtendedKeyUsage to be processed "at all" (by `obj_trust()`).  The behavior when anyExtendedKeyUsage is present (and the flag is set) depends on whether the OID is present in the 'reject' or 'trust' attributes.

> Isn't it a shame that there is so much unclarity even among experts what various verification options do?

Yes.  This lack of clarity, and the under-documented nature of many things, leads many of us to be very reluctant to introduce behavior changes...

Reply:
Yes, in other words: `obj_trust()` ignores any `anyExtendedKeyUsage` entry in the trust or reject settings unless `X509_TRUST_OK_ANY_EKU` is set.
If `X509_TRUST_OK_ANY_EKU` is set, `anyExtendedKeyUsage` in the reject list causes all purposes to be rejected (which is not properly reflected in the name  `X509_TRUST_OK_ANY_EKU`), and
if `X509_TRUST_OK_ANY_EKU` is set, `anyExtendedKeyUsage` in the trust list causes all purposes to be accepted.
This should be clearly documented.
So far, there is no documentation except for the following comment in `x509.h.in`:
```
/* Accept "anyEKU" as a wildcard trust OID */
```
As a quick initial fix I've just extended this comment as:
```
/* Accept "anyEKU" as a wildcard rejection OID and as a wildcard trust OID */
```
within #13735.
```